src/codec.rs

```diff
@@ -0,0 +1,303 @@
+            Jose::Encryption(jwe) => jwe.try_into(),
+        }
+    }
+}
+
+impl TryFrom<Encoded> for Jose {
+    type Error = Error;
+
+    fn try_from(value: Encoded) -> Result<Self, Self::Error> {
+        Ok(match value.payload {
+            Some(_) => Jose::Signature(value.try_into()?),
+            None => Jose::Encryption(value.try_into()?),
+        })
+    }
+}
+
+#[derive(PartialEq, Default, Debug, Serialize, Deserialize)]
+pub struct EncodedSignature {
+    // NOTE: Serde serialized fields in the order they appear
+    // However DAG-CBOR specifies the fields must be serialized
+    // sorted by length of key.
+    //
+    // We explicilty order the fields correctly below
+    #[serde(skip_serializing_if = "Option::is_none")]
+    header: Option<BTreeMap<String, Ipld>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    protected: Option<Bytes>,
+    signature: Bytes,
+}
+
+impl<'a> TryFrom<&'a Signature> for EncodedSignature {
+    type Error = Error;
+
+    fn try_from(value: &'a Signature) -> Result<Self, Self::Error> {
+        Ok(Self {
+            header: if value.header.is_empty() {
+                None
+            } else {
+                Some(value.header.to_owned())
+            },
+            protected: Option::from_base64(value.protected.as_ref().map(|v| v.as_str()))?,
+            signature: Bytes::from_base64(value.signature.as_str())?,
+        })
+    }
+}
+
+impl From<EncodedSignature> for Signature {
+    fn from(value: EncodedSignature) -> Self {
+        Self {
+            header: value.header.unwrap_or_else(|| BTreeMap::new()),
+            protected: value.protected.to_base64(),
+            signature: value.signature.to_base64(),
+        }
+    }
+}
+
+#[derive(PartialEq, Default, Debug, Serialize, Deserialize)]
+pub struct EncodedRecipient {
+    // NOTE: Serde serialized fields in the order they appear
+    // However DAG-CBOR specifies the fields must be serialized
+    // sorted by length of key.
+    //
+    // We explicilty order the fields correctly below
+    #[serde(skip_serializing_if = "Option::is_none")]
+    header: Option<BTreeMap<String, Ipld>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    encrypted_key: Option<Bytes>,
+}
+
+impl<'a> TryFrom<&'a Recipient> for EncodedRecipient {
+    type Error = Error;
+
+    fn try_from(value: &'a Recipient) -> Result<Self, Self::Error> {
+        Ok(Self {
+            header: if value.header.is_empty() {
+                None
+            } else {
+                Some(value.header.to_owned())
+            },
+            encrypted_key: Option::from_base64(value.encrypted_key.as_ref().map(|v| v.as_str()))?,
+        })
+    }
+}
+
+impl From<EncodedRecipient> for Recipient {
+    fn from(value: EncodedRecipient) -> Self {
+        Self {
+            encrypted_key: value.encrypted_key.to_base64(),
+            header: value.header.unwrap_or_else(|| BTreeMap::new()),
+        }
+    }
+}
+
+trait FromBase64<T>: Sized {
```

Comment:
Should match the same type bounds https://github.com/magiclen/base64-url/blob/master/src/decode.rs#L3